src/irr/asset/CGLSLVirtualTexturingBuiltinIncludeLoader.h

```diff
@@ -0,0 +1,368 @@
+	// get lower if needed (speculative execution, had to move divergent indexing to a single place)
+    vec3 loPhysCoord;
+	// speculative if (haveToDoTrilinear)
+	{
+		// now we have absolute guarantees that both LoD_high and LoD_low are in the valid original mip range
+		bool highNotInLastFull = LoD_high<originalMaxFullMip;
+		clippedLoD = highNotInLastFull ? (clippedLoD+1):clippedLoD;
+		levelInTail = highNotInLastFull ? levelInTail:(levelInTail+1);
+		loPhysCoord = irr_glsl_vTextureGrad_helper(formatID,virtualUV,clippedLoD,levelInTail);
+	}
+
+	vec4 hiMip_retval;
+    vec4 loMip;
+#ifdef IRR_GL_EXT_nonuniform_qualifier
+    hiMip_retval = textureGrad(physicalTileStorageFormatView[nonuniformEXT(formatID)],hiPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
+    if (haveToDoTrilinear)
+        loMip = textureGrad(physicalTileStorageFormatView[nonuniformEXT(formatID)],loPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
+#else
+    uint64_t outstandingSampleMask = ballotARB(true);
+    // maybe unroll a few times manually
+    while (outstandingSampleMask!=uint64_t(0u))
+    {
+		uvec2 tmp = unpackUint2x32(outstandingSampleMask);
+        uint subgroupFormatID = readInvocationARB(formatID,tmp[1]!=0u ? 32u:findLSB(tmp[0]));
+        bool canSample = subgroupFormatID==formatID; // do I need this? && (outstandingSampleMask&gl_SubGroupEqMaskARB)==gl_SubGroupEqMaskARB;
+        outstandingSampleMask ^= ballotARB(canSample);
+        if (canSample)
+        {
+            hiMip_retval = textureGrad(physicalTileStorageFormatView[subgroupFormatID],hiPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
+            if (haveToDoTrilinear)
+                loMip = textureGrad(physicalTileStorageFormatView[subgroupFormatID],loPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
+        }
+    }
+#endif
+    if (haveToDoTrilinear)
+	    hiMip_retval = mix(hiMip_retval,loMip,LoD-float(LoD_high));
+    return hiMip_retval;
+}
+
+float irr_glsl_wrapTexCoord(float tc, in uint mode)
+{
+    switch (mode)
+    {
+    case irr_glsl_STextureData_WRAP_REPEAT: tc = fract(tc); break;
+    case irr_glsl_STextureData_WRAP_CLAMP:  tc = clamp(tc, 0.0, 1.0); break;
+    case irr_glsl_STextureData_WRAP_MIRROR: tc = 1.0 - abs(mod(tc,2.0)-1.0); break;
+    default: break;
+    }
+    return tc;
+}
+vec4 irr_glsl_textureVT(in uvec2 _texData, in vec2 uv, in mat2 dUV)
+{
+    vec2 originalSz = irr_glsl_unpackSize(_texData);
+	dUV[0] *= originalSz;
+	dUV[1] *= originalSz;
+
+    uvec2 wrap = irr_glsl_unpackWrapModes(_texData);
+    uv.x = irr_glsl_wrapTexCoord(uv.x,wrap.x);
+    uv.y = irr_glsl_wrapTexCoord(uv.y,wrap.y);
+
+	vec3 virtualUV = irr_glsl_unpackVirtualUV(_texData);
+
+    uint formatID = irr_glsl_VT_layer2pid(uint(virtualUV.z));
+
+    virtualUV.xy += uv*originalSz;
+    virtualUV.xy *= irr_glsl_VT_getVTexSzRcp(formatID);
+
+    return irr_glsl_vTextureGrad(formatID, virtualUV, dUV, int(irr_glsl_unpackMaxMipInVT(_texData)));
+}
+)";
```

Comment:
you need a `ivec4 irr_glsl_vTextureLoD(in uvec2 _texData, in vec2 uv, in lod)` and `uvec4 irr_glsl_vTextureLoD(in uvec2 _texData, in vec2 uv, in lod)` as well for integer formats

Here's my stab at it:
```glsl

// textureLod emulation
ivec4 irr_glsl_vTextureLod_impl(in uint formatID, in vec3 virtualUV, in uint lod, in int originalMaxFullMip)
{
    int nonnegativeLod = int(lod);
    // magnification samples LoD 0, else clip to max representable in VT
    int clippedLoD = min(nonnegativeLod,originalMaxFullMip);
    // if minification is being performaed then get tail position
    // assert(nonnegativeLod<=originalMaxFullMip)
    int levelInTail = nonnegativeLod - clippedLoD;
    //
    vec3 physCoord = irr_glsl_vTextureGrad_helper(formatID, virtualUV, clippedLoD, levelInTail);
#ifdef IRR_GL_EXT_nonuniform_qualifier
    return textureLod(iphysicalTileStorageFormatView[nonuniformEXT(formatID)], physCoord, lod);
#else
    vec4 retval;
    uint64_t outstandingSampleMask = ballotARB(true);
    // maybe unroll a few times manually
    while (outstandingSampleMask != uint64_t(0u))
    {
        uvec2 tmp = unpackUint2x32(outstandingSampleMask);
        uint subgroupFormatID = readInvocationARB(formatID, tmp[1] != 0u ? 32u : findLSB(tmp[0]));
        bool canSample = subgroupFormatID == formatID; // do I need this? && (outstandingSampleMask&gl_SubGroupEqMaskARB)==gl_SubGroupEqMaskARB;
        outstandingSampleMask ^= ballotARB(canSample);
        if (canSample)
            retval = textureLod(iphysicalTileStorageFormatView[subgroupFormatID], physCoord, lod);
    }
    return retval;
#endif
}

ivec4 irr_glsl_vTextureLod(in uvec2 _texData, in vec2 uv, in uint lod)
{
    vec2 originalSz = irr_glsl_unpackSize(_texData);

    uvec2 wrap = irr_glsl_unpackWrapModes(_texData);
    uv.x = irr_glsl_wrapTexCoord(uv.x, wrap.x);
    uv.y = irr_glsl_wrapTexCoord(uv.y, wrap.y);
    
    vec3 virtualUV = irr_glsl_unpackVirtualUV(_texData);
    uint formatID = irr_glsl_VT_layer2pid(uint(virtualUV.z));
    virtualUV.xy += uv * originalSz;
    virtualUV.xy *= irr_glsl_VT_getVTexSzRcp(formatID);

    return irr_glsl_vTextureLod_impl(formatID, virtualUV, lod, int(irr_glsl_unpackMaxMipInVT(_texData)));
}
```